vec2text/trainers/corrector.py

```diff
@@ -305,6 +305,96 @@ def generate(
+        """
+        try:
+            frozen_embeddings = inputs["frozen_embeddings"]
+            hypothesis_input_ids = inputs["hypothesis_input_ids"]
+            hypothesis_attention_mask = inputs["hypothesis_attention_mask"]
+            hypothesis_embedding = inputs["hypothesis_embedding"]
+        except KeyError:
+            (
+                frozen_embeddings,
+                hypothesis_input_ids,
+                hypothesis_attention_mask,
+                hypothesis_embedding,
+            ) = self._get_hypothesis_uncached(inputs=inputs)
+
+        # Add beam dimension:
+        #       (batch, ...) -> (batch, beam, ...)
+        inputs["frozen_embeddings"] = frozen_embeddings
+        inputs["hypothesis_input_ids"] = hypothesis_input_ids
+        inputs["hypothesis_attention_mask"] = hypothesis_attention_mask
+        inputs["hypothesis_embedding"] = hypothesis_embedding
+
+        num_recursive_steps = num_recursive_steps or self.num_gen_recursive_steps
+        sequence_beam_width = sequence_beam_width or self.sequence_beam_width
+        num_recursive_steps_so_far = 0
+
+        total_best_scores_seen = None  # Track best scores for early stopping
+
+        ground_truth_embedding = inputs["hypothesis_embedding"]
+        hypothesis_embeddings = [ground_truth_embedding]  # Track hypothesis embeddings
+
+        hypothesis_ids = [inputs["hypothesis_input_ids"]]  # Track hypothesis ids
+
+        while num_recursive_steps >= 1:
+            gen_text_ids, hypothesis_embedding, best_scores = self._generate_with_beam(
+                inputs=inputs,
+                generation_kwargs=generation_kwargs,
+                num_recursive_steps=num_recursive_steps,
+                num_recursive_steps_so_far=num_recursive_steps_so_far,
+                sequence_beam_width=sequence_beam_width,
+            )
+            inputs["hypothesis_input_ids"] = gen_text_ids
+            inputs["hypothesis_attention_mask"] = (
+                gen_text_ids != self.model.encoder_decoder.config.pad_token_id
+            ).int()
+            inputs["hypothesis_embedding"] = hypothesis_embedding
+            # step counters
+            num_recursive_steps -= 1
+            num_recursive_steps_so_far += 1
+            # early stopping
+
+            if best_scores is not None:
+                closest_idx = torch.argmax(best_scores)
+                if (total_best_scores_seen is not None) and torch.isclose(
+                    best_scores, total_best_scores_seen, atol=1e-3
+                ):
+                    print(
+                        "scores stopped increasing! stopping early after",
+                        num_recursive_steps_so_far,
+                        "steps",
+                    )
+                    break
+                best_scores = total_best_scores_seen
+            else:
+                closest_idx = 0
+
+            # this doesn't seem to work very well?
```

Comment:
what is this comment?

Reply:
yeah just wanted to draw your attention in case you thought using `best_scores` to find the best beam was wrong. I'll remove these commented lines